src/ProgressMeter.jl

```diff
@@ -66,6 +94,7 @@ mutable struct Progress <: AbstractProgress
     tlast::Float64
     printed::Bool              # true if we have issued at least one status update
     desc::String               # prefix to the percentage, e.g.  "Computing..."
+    rate_format::AbstractRateFormat
```

Comment:
Should we introduce a type parameter to `Progress` so that `rate_format` can be concretely typed? Otherwise, the calls to `rate_string` would do dynamic dispatch. But maybe that is insignificant performance-wise compared to the IO. Could you benchmark an empty loop to see whether this PR incurs any relevant performance hit?

Reply:
Ooof. Yeah having the dynamic dispatch inside a hot loop sounds like a bad idea.

Instead of having `rate_format::AbstractRateFormat`, perhaps we could have a concretely-typed `format_string::String`?

Reply:
With the `master` branch:
```julia-repl
julia> p = ProgressMeter.Progress(10000000);

julia> p.rate_format  # master doesn't support rate format
ERROR: type Progress has no field rate_format
...

julia> @btime ProgressMeter.next!($p)
Progress: 100%|███████████████████████████████████████████████████████████████████████████████████████| Time: 0:00:02
  16.335 ns (0 allocations: 0 bytes)
```

With the present PR:
```julia-repl
julia> p = ProgressMeter.Progress(10000000);

julia> p.rate_format  # present PR supports rate format
PercentRateFormat(0)

julia> @btime ProgressMeter.next!($p)
Progress: 100%|███████████████████████████████████████████████████████████████████████████████████████| Time: 0:00:02
  15.517 ns (0 allocations: 0 bytes)
```
There doesn't seem any noticeable performance degradation.

Reply:
More informative might be something like this:
```julia
julia> p = ProgressMeter.Progress(10000000; dt=0.001);

julia> @bprofile ProgressMeter.next!($p)
Progress: 100%|█████████████████████████████████████████████████████████████████████████████████| Time: 0:00:02
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  5.342 ns … 624.207 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     6.455 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   8.209 ns ±  15.537 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▄█▄▄▂                                                       ▁
  ██████▆▅▆▇▆▆▆▅▅▅▄▅▃▃▁▄▃▁▃▄▃▃▃▃▃▁▁▁▃▃▄▁▄▁▃▁▄▃▁▃▁▃▁▁▄▁▄▅▅▄▄▄▄ █
  5.34 ns      Histogram: log(frequency) by time      55.2 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
Presumably the dispatch is only triggered when it prints?